### PR TITLE
Using a dependency graph to determine the order in which to record an oracle schema

### PR DESCRIPTION
**Pull request by croutledgeaais**

### CODE REVIEW

File: source/DataAccess/Oracle/OracleDDLHelper.ts [hunk 1]
```diff
@@ -18,8 +28,8 @@ export default class OracleDDLHelper implements IDDLHelper {
             .map((name) => {
                 // This apparently does not return metadata for some system tables/views like ALL_TABLES but will return
                 // proper DDL otherwise
-                return `SELECT 1 as "priority", TABLE_NAME, DBMS_METADATA.GET_DDL(\'TABLE\', TABLE_NAME, OWNER) as ddl ` +
-                `FROM ALL_TABLES where TABLE_NAME = \'${name}\' `;
+                return `SELECT ${prioritizedObjects.indexOf(name)} as "priority", TABLE_NAME, DBMS_METADATA.GET_DDL(\'TABLE\', TABLE_NAME, OWNER) as ddl ` +
```
Comment (croutledgeaais) [inline source/DataAccess/Oracle/OracleDDLHelper.ts]: Repurposing the 'priority' of each record. At a later time indexes can appear well after all tables.

File: source/DataAccess/Oracle/OracleDDLHelper.ts [hunk 2]
```diff
@@ -28,6 +38,49 @@ export default class OracleDDLHelper implements IDDLHelper {
...
+                JOIN ALL_CONSTRAINTS cons
+                ON cons.R_OWNER = col.OWNER
+                AND cons.R_CONSTRAINT_NAME = col.CONSTRAINT_NAME
+            WHERE cons.TABLE_NAME = '${tableOrViewName}'`;
+        })
+        .join('\nUNION ALL\n');
+
+        const records = await connection.execute(tableConstraintQueries);
+
+        const tableGraph = new DepGraph<string>();
```
Comment (croutledgeaais) [inline source/DataAccess/Oracle/OracleDDLHelper.ts]: A handy dependency graph class makes this task easy!

File: source/DataAccess/Oracle/OracleDDLHelper.ts [hunk 3]
```diff
@@ -28,6 +38,49 @@ export default class OracleDDLHelper implements IDDLHelper {
...
+        validTableNames.forEach((tableOrView, idx, array) => tableGraph.addNode(tableOrView));
+
+        // Use a simple 2-tuple to represent an association of two database objects
+        let row: [string, string];
+        for (row of records.rows) {
+            // We don't care about self referencing tables so only worry about
+            // inter-table dependencies
+            if (row[0] !== row[1]) {
```
Comment (croutledgeaais) [inline source/DataAccess/Oracle/OracleDDLHelper.ts]: Only when a series of tables creates a circular dependency will we fail to report the correct list of prioritized records. The DepGraph class is also explicitly being created without the circular: true flag and so will fail when there's a nasty non-self referencing circular dependency. We'll need to address that once we find one in the wild.

File: test/unit/OracleDDLHelper.spec.ts [hunk 4]
```diff
@@ -1,23 +1,105 @@
...
+chai.use(chaiAsPromised);
 
 import OracleDDLHelper from '../../source/DataAccess/Oracle/OracleDDLHelper';
+import IConnectionPool from '../../source/DataAccess/IConnectionPool';
 
 const expect = chai.expect;
 
 describe('OracleDDLHelper', () => {
 
+    const createTestContext = (tableAssociations: Array<[string, string]> = []) => {
```
Comment (croutledgeaais) [inline test/unit/OracleDDLHelper.spec.ts]: I got fed up with the caveats that come with trying to stub layers of stubbed objects and just went with anonymous stubs.

File: test/unit/OracleDDLHelper.spec.ts [hunk 5]
```diff
@@ -1,23 +1,105 @@
...
+
+            expect(query).to.contain('SELECT 0 as "priority", TABLE_NAME, DBMS_METADATA.GET_DDL(\'TABLE\', TABLE_NAME, OWNER) as ddl FROM ALL_TABLES WHERE TABLE_NAME = \'table2\'');
+            expect(query).to.contain('SELECT 1 as "priority", TABLE_NAME, DBMS_METADATA.GET_DDL(\'TABLE\', TABLE_NAME, OWNER) as ddl FROM ALL_TABLES WHERE TABLE_NAME = \'table1\'');
+            expect(query).to.contain('SELECT 2 as "priority", TABLE_NAME, DBMS_METADATA.GET_DDL(\'TABLE\', TABLE_NAME, OWNER) as ddl FROM ALL_TABLES WHERE TABLE_NAME = \'table3\'');
+        });
+
+        it('should return a properly formatted query for multiple related and unrelated tables', async () => {
+            const ctx = createTestContext([['table1', 'table2']]);
+            const helper = new OracleDDLHelper(ctx.pool);
+            const query = await helper.getDDLQuery(['table3', 'table1', 'table2']);
```
Comment (croutledgeaais) [inline test/unit/OracleDDLHelper.spec.ts]: In cases like this it really doesn't matter where table3 lands in the priority list as there are no dependencies on it. Placing  it at the end of the 'valid table names' arg here would land it at the end of the prioritized list without any side effects.

File: source/DataAccess/Oracle/OracleDDLHelper.ts (hunk 3 above)
Comment (aregier) [reply source/DataAccess/Oracle/OracleDDLHelper.ts]: AFAIK, the only way to create a circular in RDBMS is to make a nullable foreign key. I'm unfortunate enough to have seen it used in software design, but it's pretty unlikely we would see a schema that does this.

**Pull request merged**